pkg/ipcache/ipcache.go

```diff
@@ -653,8 +669,24 @@ func (ipc *IPCache) deleteLocked(ip string, source source.Source) (namedPortsCha
 		namedPortsChanged = namedPortsChanged && ipc.needNamedPorts.Load()
 	}
 
-	if newHostIP != nil {
-		nodeID = ipc.AllocateNodeID(newHostIP)
+	// If this is a straight delete this is simple, deallocate the nodeID. If its a
+	// Upsert then the deleted IP was shadowing another CIDR so we will restore
+	// that mapping. But, from NodeID side we already did an allocate for that and
+	// we don't want to bump the refcnt here so simply Get() the ID. Remember to
+	// delete the old IP through the Upsert path though.
+	if oldHostIP != nil {
+		if cacheModification == Delete {
+			ipc.DeallocateNodeID(oldHostIP)
+		} else if cacheModification == Upsert {
+			if oldHostIP.Equal(newHostIP) {
+				nodeID, _ = ipc.GetNodeID(oldHostIP)
+			} else {
+				if newHostIP != nil {
+					nodeID, _ = ipc.GetNodeID(newHostIP)
```

Comment:
I'll need to double check... I thought in this case the newHostIP already exists with a nodeID so we don't want to inc it again because we will only get a single delete for it. I'll double check and clarify shortly.

Reply:
I'm not confident so maybe need an ipcache expert here? Here is my reasoning.

This is the 'cacheModification = Upsert' case  with newHostIP != oldHostIP.  So we got here through this,

```
                // Check whether the deleted endpoint IP was shadowing that CIDR, and                                                                                                                                                                                          
                // restore its mapping with the listeners if that was the case.                                                                                                                                                 
                cidrClusterStr := cidrCluster.String()                                                                                                                                                                                                                                        
                if cidrIdentity, cidrFound := ipc.ipToIdentityCache[cidrClusterStr]; cidrFound {                                                                                                                                                
                        newHostIP, _ = ipc.getHostIPCache(cidrClusterStr)                                                                                                                                                       
                        if cidrIdentity.ID != cachedIdentity.ID || !oldHostIP.Equal(newHostIP) {                                                                                                                                              
                                scopedLog.Debug("Removal of endpoint IP revives shadowed CIDR to identity mapping")                                                                                                                                                                          
                                cacheModification = Upsert                                                                                                                                                                                                                                   
                                cidrIdentity.shadowed = false                                                                                                                                                                                         
                                ipc.ipToIdentityCache[cidrClusterStr] = cidrIdentity                                                                                                                                                                                             
                                oldIdentity = &cachedIdentity                                                                                                                                                                                                                       
                                newIdentity = cidrIdentity          
```
so we got the newHostIP with getHostIPCache which is returning a mapping already in the ipcache? So if we already put it in the ipcache then we don't need to allocate for it because its already been refcnt'd by the thing that put it there. And then when its removed it will either get a nodeDelete or ipcache delete to decrement the refcnt. I'm pretty sure this lines up with experimentation where having an alloc here (which is what I originally did) got some hostIDs stuck, but would need to double check that to be sure.